app/collection.go

```diff
@@ -7,37 +7,49 @@ import (
 	"net/http"
 
 	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/utils"
 )
 
 func (a *App) registerCollectionAndTopic(pluginID, collectionType, topicType string) error {
-	collectionTypeForPlugin, ok := a.ch.collectionTypes[pluginID]
-	if ok && collectionTypeForPlugin != collectionType {
+	// we have a race condition due to multiple plugins calling this method
+	a.ch.collectionMut.Lock()
+	defer a.ch.collectionMut.Unlock()
+
+	// check if collectionType was already registered by other plugin
+	if pluginIDForCollection, found := getKeyWithValue(a.ch.collectionTypes, collectionType); found && pluginIDForCollection != pluginID {
 		return model.NewAppError("registerCollectionAndTopic", "app.collection.add_collection.exists.app_error", nil, "", http.StatusBadRequest)
-	} else if !ok {
-		a.ch.collectionTypes[pluginID] = collectionType
 	}
 
-	for _, ts := range a.ch.topicTypes {
-		if containsSlice(ts, topicType) {
+	// check if topicType was already registered to other collection
+	for collectionTypeForTopic, topicTypes := range a.ch.topicTypes {
+		if utils.StringInSlice(topicType, topicTypes) && collectionTypeForTopic != collectionType {
 			return model.NewAppError("registerCollectionAndTopic", "app.collection.add_topic.exists.app_error", nil, "", http.StatusBadRequest)
 		}
 	}
 
-	if ts, ok := a.ch.topicTypes[collectionType]; !ok {
-		a.ch.topicTypes[collectionType] = []string{topicType}
+	collectionTypesForPlugin, ok := a.ch.collectionTypes[pluginID]
+	if ok && !utils.StringInSlice(collectionType, collectionTypesForPlugin) {
+		collectionTypesForPlugin = append(collectionTypesForPlugin, collectionType)
+		a.ch.collectionTypes[pluginID] = collectionTypesForPlugin
+	} else if !ok {
+		a.ch.collectionTypes[pluginID] = []string{collectionType}
+	}
```

Comment:
Since slices don't need initialization (unlike maps), is it safe to reduce this to just the following?
```suggestion
		a.ch.collectionTypes[pluginID] = append(a.ch.collectionTypes[pluginID], collectionType)
```